feat(perf): Add tags page prototype backend

k-fish opened the pull request:
### Summary
This adds a prototype backend for the upcoming tags page, extending the existing endpoint with params useful for the tag page, and adding a new endpoint for performance facet histograms for a tag key.

#### Other
- This modifies histogram_query to accept an optional group_by, in addition to the filtered grouping. I'm up for counter-suggestions.

wmak commented on src/sentry/api/endpoints/organization_events_facets_performance.py:
```diff
@@ -30,6 +35,19 @@ def get(self, request, organization):
+        histograms = None
+
+        if self.has_tag_page_feature(organization, request):
+            # TODO(k-fish): Temporarily allow these options for this feature, potentially circle back to split histogram off before release.
```
by this do you mean split histograms to its own endpoint?

k-fish replied:
Not sure, thoughts? I personally prefer separate endpoints for different data structures. What did you have in mind?

wmak replied:
yeah IMO we shouldn't change the response structure of this endpoint based on params. so either a param for the existing histogram endpoint or a completely separate one

dashed replied:
A separate endpoint would be amazing.

wmak commented on src/sentry/snuba/discover.py:
```diff
@@ -848,16 +858,32 @@ def histogram_query(
-        functions_acl=["array_join", "histogram"],
-    )
+
+    if group_by:
+        results = query(
+            selected_columns=columns + [histogram_column, "count()"],
+            conditions=conditions,
+            query=user_query,
+            params=params,
+            orderby=[histogram_alias],
+            limit=len(fields) * num_buckets,
+            group_by=group_by,
+            referrer=referrer,
+            functions_acl=["array_join", "histogram"],
+        )
+        return results
+
+    else:
+        results = query(
+            selected_columns=columns + [histogram_column, "count()"],
+            conditions=conditions,
+            query=user_query,
+            params=params,
+            orderby=[histogram_alias],
+            limit=len(fields) * num_buckets,
+            referrer=referrer,
+            functions_acl=["array_join", "histogram"],
+        )
```
Instead could we refactor this to use `prepare_discover_query` instead? something like
```python
histogram_query = prepare_discover_query(...)
if group_by:
    histogram_query.groupby += group_by
 ```

k-fish committed:
feat(perf): Add tags page prototype backend
This adds a prototype backend for the upcoming tags page, and currently uses params passed to the existing facets endpoint to serve different results, although this can be separate endpoints in the future.

k-fish committed:
Switch to using prepare_discover_query

k-fish committed:
Add new endpoint solely for tag performance facet histograms

k-fish committed:
Remove histograms from existing endpoint

k-fish committed:
Merge remote-tracking branch 'origin/master' into feat/perf-tags-subpage-backend

wmak reviewed (approved):
IMO we should put both endpoints in the same file, I'd even have them inherit from a shared base class since they share some common features like project length validation, and `query_tag_data`

k-fish committed:
Refactor both endpoints into one file

k-fish committed:
Merge remote-tracking branch 'origin/master' into feat/perf-tags-subpage-backend

k-fish committed:
Clean up the setup function to have a single return type and raise errors instead

k-fish committed:
Remove code post-return
